File: scripts/download-moon-images.js
const fs = require('fs');
const https = require('https');
const path = require('path');

const BASE_URL = 'https://svs.gsfc.nasa.gov/vis/a000000/a004600/a004605/frames/730x730_1x1_30p/';
const TOTAL_IMAGES = 8761;
const OUTPUT_DIR = path.join(process.cwd(), 'public', 'moon-phases');

// Create output directory if it doesn't exist
if (!fs.existsSync(OUTPUT_DIR)) {
  fs.mkdirSync(OUTPUT_DIR, { recursive: true });
}

// Find the last downloaded image
function findLastDownloadedImage() {
  const files = fs.readdirSync(OUTPUT_DIR);
  if (files.length === 0) return 0;
  
  const numbers = files
    .map(f => parseInt(f.match(/moon\.(\d+)\.jpg/)?.[1] || '0'))
    .filter(n => !isNaN(n));
  
  return Math.max(0, ...numbers);
}

async function downloadImage(imageNumber, retries = 3) {
  const paddedNumber = String(imageNumber).padStart(4, '0');
  const fileName = `moon.${paddedNumber}.jpg`;
  const url = BASE_URL + fileName;
  const outputPath = path.join(OUTPUT_DIR, fileName);

  return new Promise((resolve, reject) => {
    const download = () => {
      https.get(url, (response) => {
        if (response.statusCode !== 200) {
          if (retries > 0) {
            console.log(`Retrying ${fileName} (${retries} attempts left)`);
            setTimeout(() => {
              downloadImage(imageNumber, retries - 1)
                .then(resolve)
                .catch(reject);
            }, 1000);
            return;
          }
          reject(new Error(`Failed to download ${fileName}`));
          return;
        }

        const fileStream = fs.createWriteStream(outputPath);
        response.pipe(fileStream);

        fileStream.on('finish', () => {
          fileStream.close();
          console.log(`Downloaded ${fileName}`);
          resolve();
        });
      }).on('error', (err) => {
        if (retries > 0) {
          console.log(`Retrying ${fileName} (${retries} attempts left)`);
          setTimeout(() => {
            downloadImage(imageNumber, retries - 1)
              .then(resolve)
              .catch(reject);
          }, 1000);
        } else {
          reject(err);
        }
      });
    };

    download();
  });
}

async function downloadAllImages() {
  const startFrom = findLastDownloadedImage() + 1;
  console.log(`Starting download from image ${startFrom} of ${TOTAL_IMAGES}...`);
  
  // Download images in batches to avoid overwhelming the server
  const BATCH_SIZE = 5;
  
  for (let i = startFrom; i <= TOTAL_IMAGES; i += BATCH_SIZE) {
    const batch = [];
    for (let j = i; j < Math.min(i + BATCH_SIZE, TOTAL_IMAGES + 1); j++) {
      batch.push(downloadImage(j));
    }
    try {
      await Promise.all(batch);
      console.log(`Completed batch ${Math.floor(i/BATCH_SIZE) + 1} of ${Math.ceil(TOTAL_IMAGES/BATCH_SIZE)}`);
    } catch (error) {
      console.error('Error in batch:', error);
      // Continue with next batch even if current batch had errors
    }
  }

  console.log('All images downloaded successfully!');
}

downloadAllImages().catch(console.error); 

File: scripts/download_moon_images.py
import os
import requests
import time
from concurrent.futures import ThreadPoolExecutor
from pathlib import Path

# Create output directory
output_dir = Path("public/moon-phases")
output_dir.mkdir(parents=True, exist_ok=True)

def download_image(hour):
    """Download a single moon phase image for the given hour."""
    # NASA's image URL format
    url = f"https://svs.gsfc.nasa.gov/vis/a000000/a004500/a004537/frames/730x730_1x1_30p/moon.{hour:04d}.jpg"
    output_path = output_dir / f"moon.{hour:04d}.jpg"
    
    # Skip if file exists and has content
    if output_path.exists() and output_path.stat().st_size > 0:
        print(f"Skipping existing file: {output_path}")
        return
    
    try:
        response = requests.get(url, timeout=30)
        response.raise_for_status()
        
        with open(output_path, 'wb') as f:
            f.write(response.content)
        print(f"Downloaded: {output_path}")
        
        # Be nice to NASA's servers
        time.sleep(0.5)
        
    except Exception as e:
        print(f"Error downloading hour {hour}: {e}")

def main():
    # Start from 4216 and continue to 8760
    start_hour = 4216
    total_hours = 8760
    
    print(f"Starting download from hour {start_hour} to {total_hours}...")
    
    # Use thread pool to speed up downloads while being respectful
    with ThreadPoolExecutor(max_workers=4) as executor:
        executor.map(download_image, range(start_hour, total_hours + 1))
    
    print("Download complete!")

if __name__ == "__main__":
    main() 